linkerd/app/core/src/svc.rs

```diff
@@ -233,6 +235,47 @@ impl<S> Stack<S> {
         }))
     }
 
+    pub fn push_discover_cache<D, K, T>(
+        self,
+        discover: D,
+        idle: Duration,
+    ) -> Stack<
+        crate::disco_cache::NewDiscoveryCache<
+            K,
+            impl NewService<
+                    K,
+                    Service = impl Service<
+                        (),
+                        Error = Error,
+                        Response = D::Response,
+                        Future = impl Send + 'static,
+                    > + Clone
+                                  + Send
+                                  + Sync
+                                  + 'static,
+                > + Clone
+                + Send
+                + Sync
+                + 'static,
+            S,
+        >,
+    >
+    where
+        T: Param<K> + Clone,
+        K: Clone + fmt::Debug + Eq + Hash + Send + Sync + 'static,
+        D: Service<K> + Clone + Send + Sync + 'static,
+        D::Error: Into<Error> + Send + 'static,
+        D::Future: Send + 'static,
+        D::Response: Clone + Send + Sync + 'static,
+        S: NewService<(D::Response, T)> + Clone,
+    {
+        let discover = stack(discover)
+            .push(stack::SpawnThunkCloneResponse::layer())
+            .check_new_service::<K, ()>();
+        self.push(crate::disco_cache::NewDiscoveryCache::layer(discover, idle))
```

Comment:
Can we use `NewDiscoveryCache` without `SpawnThunkCloseResponse`? If not, it's probably simpler to move it into the disco_cache module. Specifically it eliminates the awful type mess.